File: 2019W17/PreppindataW17.py
import pandas as pd

vote = pd.read_csv("E:\Voting Systems.csv", sep='\t')
winner = {}

#FPTP    
vote['Rank1'] = vote.apply(lambda x: list(x['Voting Preferences'])[0],axis=1)
winner['FPTP']=vote.groupby(['Rank1']).count().Voter.idxmax()

#Borda Count
vote['Rank2'] = vote.apply(lambda x: list(x['Voting Preferences'])[1],axis=1)
vote['Rank3'] = vote.apply(lambda x: list(x['Voting Preferences'])[2],axis=1)
bcvote = vote.melt(id_vars=['Voting Preferences','Voter'], var_name='ranking', value_name='candidate')[['candidate','ranking']]
bcvote['point'] = bcvote.apply(lambda x: 4-int(list(x['ranking'])[-1]),axis=1)
winner['Borda Count']=bcvote.groupby(['candidate']).sum().point.idxmax()

#AV
remove =[]
while True: 
    vote['AVcandidate'] = vote.apply(lambda x: [c for c in list(x['Voting Preferences']) if c not in remove][0],axis=1)
    if max(vote.groupby(['AVcandidate']).count().Voter)>len(vote)*0.5:
        winner['AV']=vote.groupby(['AVcandidate']).count().Voter.idxmax()
        break
    else:
        remove.append(vote.groupby(['AVcandidate']).count().Voter.idxmin())
        
#Final
final = pd.DataFrame(winner.items(), columns=['Voting System', 'Winner'])    

File: 2019/2019W17/preppindataw17.R
library(dplyr)
library(tidyr)
library(readr)

vote <- read_delim("E:/Voting Systems.csv", 
                    "\t", escape_double = FALSE, trim_ws = TRUE)

fptp <- vote %>%
  mutate('Rank1'=sapply(strsplit(`Voting Preferences`, ''), function(x) x[[1]][1])) %>%
  mutate('winner' = names(which.max(table(`Rank1`)))) %>%
  select('winner') %>%
  distinct() %>%
  mutate('Vote System' = 'fptp') 

borda_count <- vote %>%
  mutate('Rank1'=sapply(strsplit(`Voting Preferences`, ''), function(x) x[[1]][1])) %>%
  mutate('Rank2'=sapply(strsplit(`Voting Preferences`, ''), function(x) x[[2]][1])) %>%
  mutate('Rank3'=sapply(strsplit(`Voting Preferences`, ''), function(x) x[[3]][1])) %>%
  gather(., key='ranking', value='candidate', -c(`Voting Preferences`,`Voter`)) %>%
  mutate('Point'=case_when(`ranking`=='Rank1' ~ 3, `ranking`=='Rank2' ~ 2, TRUE ~ 1)) %>%
  group_by(`candidate`) %>%
  summarise('Point'=sum(`Point`)) %>%
  filter(`Point` == max(`Point`)) %>%
  mutate('winner'= `candidate`) %>%
  select('winner') %>%
  mutate('Vote System' = 'borda count') 

avtemp <- vote
avrule <- function(x) {
  avtemp <<- avtemp %>% 
    mutate('Voting Preferences'=sub(x,'',`Voting Preferences`),
           '50%Level'=n()/2) %>%
    mutate('Rank1'=sapply(strsplit(`Voting Preferences`, ''), function(x) x[[1]][1]))

  avmax <- avtemp %>%
    group_by(`Rank1`) %>%
    summarise('Count'=n(),'50%Level'=first(`50%Level`)) %>%
    filter(`Count`> `50%Level`) %>%
    select('Rank1') %>%
    rename('winner' = 'Rank1')
  
  if(nrow(avmax)==0){
    avrule(unlist(avtemp %>%
                    mutate('minvote' = names(which.min(table(`Rank1`)))) %>%
                    select('minvote') %>%
                    distinct())[[1]])
  } else {
    return(avmax)
  }
}

av <- avrule('') %>%
  mutate('Vote System' = 'av') 
  
final <- do.call('rbind', list(fptp, borda_count, av)) %>%
  select(c('Vote System','winner'))

View(final)
